### ISSUE TITLE
Multi Cluster App: CLI

### ISSUE THREAD

**Issue by alena1108** [rancher/rancher#17147]
This ticket is to track CLI work for Multi Cluster App feature: #16016

We have to add an ability to deploy and upgrade the app via rancher cli

**Comment by gitlawr** [comment]
```
$ rancher app -h
Operations with apps

Usage:
   rancher apps command [command options] [arguments...]

Commands:
     ls                  List apps
     delete              Delete an app
     install             Install an app template
     rollback            Rollback an app to a previous version
     upgrade             Upgrade an app to a newer version
     list-templates, lt  List templates available for installation
     show-template, st   Show versions available to install for an app template
     show-app, sa        Show an app's available versions and revisions

Options:
   --format value, -o value  'json', 'yaml' or custom format
   --quiet, -q               Only display IDs
   --help, -h                show help
```

The above are the utilities of existing app command. The operations for multi-cluster apps will much likely share the same set. `apps` & `multi-cluster apps` are two kinds of api objects, but it looks easier to use if we bind it to existing app command.

To do that:
Add `--multi-cluster, -m` options to all app comands. when this flag is enabled, use multi-cluster apps as the operation targets.
Add `--target` options to `install/upgrade` command

**Comment by alena1108** [comment]
@gitlawr what about answers section? It is quite different for multi cluster app. With multi cluster app it can be defined on the global, cluster, project level.

**Comment by gitlawr** [comment]
Yes, that's different. We claim that answers are sets of key value pairs in docs and helper texts which is changed in multi-cluster apps API(kv pairs+scope).
I propose to keep consistency with exsting answers file format for user' interaction with cli:
```
--answers value, -a value    Path to an answers file, the format of the file is a map with key:value. This supports JSON and YAML. #which keeps the same
--set Set answers for the template, can be used multiple times. \
Example: --set foo=bar, --set clustera:foo=bar, --set clustera:projectb:foo=bar # we extend the answer format to support scoped answers
```
In this way, the answers file remains its format:
```
defaultImage: true
images.server.repository: example/image
images.server.tag: prod
"clustera:images.server.tag": dev
"clustera:projectb:images.server.tag": test
```
We need to quote the scoped keys though so as to work with yaml spec.

**Comment by gitlawr** [comment]
I also wonder if we can use this schema in API perspective

**Comment by gitlawr** [comment]
See example command flows in https://gist.github.com/gitlawr/8fe61c24e72b9cb405bb18fb1ee03506

**Comment by cjellick** [comment]
not resolved until Dan LGTMs. Also, I believe @loganhz was going to do a review and sync on this PR as well

**Comment by cjellick** [comment]
@gitlawr dont remove status labels as an issue moves forward in the process. Status labels should just accumulate. We only remove status labels when an issue goes backwards in the flow

**Comment by gitlawr** [comment]
Need to address RBAC changes

**Comment by cjellick** [comment]
@moelsayed to validate

**Comment by alena1108** [comment]
@mrajashree to validate

**Comment by mrajashree** [comment]
@gitlawr Multiclusterapp targets is a required field in the api. Can we make it required in cli too? We have this as an example command but I think we should error out
```
./rancher multiclusterapp install redis appfoo
```

**Comment by gitlawr** [comment]
@mrajashree If no target flag is passed, the project of the current context will be used as the target, does it make sense for you?

The error shows that it's a timeout waiting for the app to be ready, the default timeout is 60 seconds, you can set it via `--timeout` as the install option. Or we might want to increase the default value?

**Comment by kinarashah** [comment]
@kinarashah to validate

**Comment by mrajashree** [comment]
Validated these commands
```
1. Creation
./rancher multiclusterapps install --target c-xx7q5:p-dhckk --target c-tfq9t:p-6tldl --role project-member grafana test1-grafana
2. Updating a role
./rancher multiclusterapps upgrade --role cluster-owner test1-grafana 0.0.31
3. Add project
./rancher multiclusterapps add-project test-graf-ap c-tfq9t:p-6tldl
4. Remove project
./rancher multiclusterapps delete-project test-graf-ap c-xx7q5:p-dhckk
```
@kinarashah is taking over and I'll focus on rbac and roles related commands

**Comment by kinarashah** [comment]
Validation commands based on test plan, 

- Deploy to single target 
`./rancher multiclusterapps install --role cluster-owner grafana mcapp` 

- Deploy to multiple targets 
`./rancher multiclusterapps install --target c-fq2pd:p-qjct4 --target c-lqqgh:p-r2pgk --role cluster-owner grafana multi` 

- Deploy with specific version 
`./rancher multiclusterapps install --target c-fq2pd:p-qjct4 --target c-lqqgh:p-r2pgk --role cluster-owner --version 1.4.3 docker-registry test2` 

- Deploy with no-prompt, suppresses answers 
` ./rancher multiclusterapps install --target c-lqqgh:p-r2p  docker-registry test3 --no-prompt` 

- Invalid targets error out 
`FATA[0000] Bad response statusCode [409]. Status [409 Conflict]. Body: [baseType=error, code=AlreadyExists, message=multiclusterapps.management.cattle.io "test2" already exists] from [https://ec2-52-27-46-128.us-west-2.compute.amazonaws.com/v3/multiclusterapps]` 
`FATA[0003] Not found: c-lqqgh:p-r2p` 

- Duplicate projects error out
`FATA[0004] Bad response statusCode [422]. Status [422 Unprocessable Entity]. Body: [baseType=error, code=InvalidBodyContent, message=duplicate projects in targets c-lqqgh:p-r2pgk] from [https://ec2-52-27-46-128.us-west-2.compute.amazonaws.com/v3/multiclusterapps]`

- List multicluster apps 
`./rancher multiclusterapps ls (-q)`
`./rancher multiclusterapps ls --format yaml` 
`./rancher multiclusterapps ls --format json` 

- Add project (with/without answers) 
`./rancher multiclusterapps add-project --set c-lqqgh:p-r2pgk:foo=bar mcapp c-lqqgh:p-r2pgk` (new ans) 
`./rancher multiclusterapps add-project --set c-lqqgh:p-r2pgk:auth.anonymous.enabled=false mcapp c-lqqgh:p-r2pgk` (existing answer) 
`./rancher multiclusterapps add-project --set c-lqqgh:p-r2pgk:auth.anonymous.enabled=false mcapp c-lqqgh:p-r2pgk c-fq2pd:p-qjct4` (multiple projects) 

- Delete project 
`./rancher multiclusterapps delete-project mcapp c-lqqgh:p-r2pgk c-fq2pd:p-qjct4` 

- Upgrade show versions 
`./rancher multiclusterapps upgrade --show-versions test2` 

- Upgrade changing template version 
`./rancher multiclusterapps upgrade test2 1.6.1` 

- Upgrade changing global answers 
`./rancher multiclusterapps upgrade --set secrets.htpasswd:"kinara" test 1.6.1`

- Upgrade setting cluster scoped answers 
`./rancher multiclusterapps upgrade --set c-lqqgh:foo=bar test2 1.6.1`

- Upgrade setting project scoped answers 
`./rancher multiclusterapps upgrade --set c-fq2pd:p-qjct4:foo=bar test3 1.6.1`

- Rollback show revisions 
`./rancher multiclusterapps rollback --show-revisions test2` 

- Rollback to a revision 
`./rancher multiclusterapps rollback test2  mcapprevision-gs48v`

**Comment by kinarashah** [comment]
Questions @gitlawr 

- I think we should have something like `show-answers` similar to `show-versions`, otherwise user needs to do `ls multiclusterapps` and then find answers.

- Upgrading just the answers doesn't work for me, am I missing something? 
`./rancher multiclusterapps upgrade --set adminPassword=bar2 multi` This is allowed from UI/API and creates a new multiclusterapp revision. 

- How to upgrade with timeout? Since this is present on the mcapp object itself, we should allow editing the object to pass upgrade strategy, interval and save it. Don't see that option under upgrade either `./rancher multiclusterapps upgrade --help`. 

- Upgrading invalid format didn't error out as expected and went through.  We should error out, 
`./rancher multiclusterapps upgrade test2 docker-registry 1.6.1` 
Trying to use UI for see multi cluster app page crashed.
```NotFound (404)
catalogtemplateversions.management.cattle.io "library-docker-registry-docker-registry" not found```

- Upgrade/Rollback doesn't wait for multi cluster app to be active like install, is this expected? 

- Should automatically allow rolling back to previous revision? 

cc @mrajashree

**Comment by gitlawr** [comment]
2. You should specify template version on upgrade, even it is not changed
```
rancher multiclusterapps upgrade --set adminPassword=bar2 multi [version]
```

The others are action items:
1. Add `show-answers` command
2. Add upgrade strategy options
3. Error out on invalid version
4. Wait for mcapp `installed` state on upgrade/rollback and add `timeout` option
5. Add `--rollback` option for upgrade to support rollback on failure

**Comment by cjellick** [comment]
remaining requests from @kinarashah 's comment are in v2.2.0-rc14

**Comment by kinarashah** [comment]
Thank you @cjellick, will test them out today

**Comment by kinarashah** [comment]
@gitlawr Thanks for fixing! 

Following things work as expected, 
- `list-answers`, 
- validation on upgrade
- upgrading with answers work as expected. 
- waiting for state to be active for upgrade/rollback 

1. For upgrading with strategy, the strategy actually is a part of the mcapp object. Just passing through cli doesn't work since it doesn't update the object. 
when I do, all apps are upgraded at once. 
`multiclusterapp upgrade --set auth.anonymous.enabled=true multi 0.0.31 --upgrade-batch-size=1 --upgrade-batch-interval=30` 
(also, since these fields are tied to upgrade, not sure if they need to be named as --upgrade-batch-size) 

2. For rollback, I see that we changed the description to `Rollback to previous revision when the upgrade fails` but I still need to pass the revision name, is that expected? 

Let me know if you want me to create a separate issue just for upgrade and close this feature one.

**Comment by gitlawr** [comment]
@kinarashah 
1. To update the batch-size and interval the `--upgrade-strategy rolling-update` is needed
2. They are named as `--upgrade-batch-size/interval` so that it is consistent with the flags for the install command.
3. The `--rollback` is a bool flag so no revision is needed for it, is there anything I'm missing?

**Comment by kinarashah** [comment]
@gitlawr 

Discussed 3 with @sangeethah, we shouldn't automatically rollback to previous one when upgrade fails, so can just remove `--rollback` during upgrade. 
Also, she suggested we shouldn't rollback to previous revision on rollback as well, to keep it consistent with API.

**Comment by loganhz** [comment]
@kinarashah v2.2.0-rc16 is ready for verification

**Comment by kinarashah** [comment]
Thank you @loganhz @gitlawr 

Confirmed that `--rollback` isn't available anymore with upgrade command in v2.2.0-rc16. 
Other upgrade commands work as expected.
